### ISSUE TITLE
CredSSP not fully working

### ISSUE THREAD

**Issue by moobyfr** [rdesktop/rdesktop#197]
Hi
My setup is:

- `2K12` server in a domain `SUB1.AD.DOMAIN.COM`
- linux used to auth users in `AD.DOMAIN.COM`
- the user is able to get a ticket with `kinit` 
When I make a RDP connexoin with rdesktop, I get a TGS for 
`TERMSRV/2K12.SUB1.AD.DOMAIN.COM@AD.DOMAIN.COM`
(`Connection established using CredSSP`)

on the server side, the auth seems to start, but it stops with "invalid password"
And the username prefilled is "AD.DOMAIN.COM\username"
this is strange, the normal login should be `AD\username` or `username@AD.DOMAIN.COM`

**Comment by derfian** [comment]
It took me a little while to get there but I was able to reproduce this problem. The steps required to reproduce this scenario for me was to get a user ticket with kinit, run rdesktop but do *not* enter a password when prompted. Also, I wasn't able to reproduce if I got redirected to a session on another RDS server.

I seem to recall somewhere reading that CredSSP + Kerberos isn't used for the user authentication, just for authenticating the server, and that you still need a password (or smart card) to authenticate. I'd love to have someone else confirm that.

Which version of rdesktop are you using?

**Comment by moobyfr** [comment]
I'm using 1.8.3 from debian repository stretch

**Comment by uglym8** [comment]
I'm not an CredSSP expert but I think @derfian is talking about the following question by @hean01-cendio :https://social.msdn.microsoft.com/Forums/SqlServer/en-US/da074f0f-0887-4151-88ea-19a671ed91d9/is-it-possible-to-do-true-kerberos-sso-using-rdp-?forum=os_windowsprotocols

Also check #71, #72

**Comment by hean01-cendio** [comment]
@uglym8 yes, that is true, CredSSP use of Kerberos is only for confidentiality and integrity to prevent MITM.

To simplify that, one should think 'is it safe to share my credentials with the remote end', kerberos ensures that.

**Comment by hean01-cendio** [comment]
@moobyfr I agree, that it is strange the REALM is used as domain name prefix and not using a user principle name eg. user@REALM. I'll look though code to see if I can spot an error upon this topic.

What is the command line for your rdesktop connection, when this fails ?

**Comment by moobyfr** [comment]
simply ``rdesktop tseserver`` while having a valid kerberos ticket

**Comment by hean01-cendio** [comment]
> simply rdesktop tseserver while having a valid kerberos ticket

Ah ok, that does not provide any domain or authentication credentials.

@moobyfr can you connect correctly when you provide the correct domain and authentication credentials via command line ? Also verify that CredSSP is used for the connection via console
output.

**Comment by trentasis** [comment]
What are the steps to configure use of credssp with rdesktop, on os side ans rdesktop options?

**Comment by trentasis** [comment]
Anybody can help?

**Comment by hean01-cendio** [comment]
@trentasis, there are information on issue #71 how to setup and use CredSSP with rdesktop.

For these general questions please do use our user mailing lists [1] instead of hijacking non related issue.

[1] https://groups.google.com/forum/#!forum/rdesktop-users

**Comment by moobyfr** [comment]
@hean01-cendio : I can log in, if I provide domain and credentials on the command line for the following case (and CredSSP is used):
 * TGT for userAD@AD.DOMAIN.COM
 * rdesktop -u userAD -p pass -d AD 2K12.SUB1.AD.DOMAIN.COM

If I try to log in with a account from the domain SUB1:
 * TGT for userAD@AD.DOMAIN.COM
 * rdesktop -u userSUB1 -p passs -d SUB1 2K12.SUB1.AD.DOMAIN.COM
it doesn't work : CredSSP is ok (I can see a correct TGS), but I get a invalid password, and I see that the used login was: AD.DOMAIN.COM\userSUB1

Il I get a TGT for SUB1:
 * TGT for userSUB1@SUB1.AD.DOMAIN.COM
 * rdesktop -u userSUB1 -p passs -d SUB1 2K12.SUB1.AD.DOMAIN.COM
It works: CredSSP is ok and The login is done.

**Comment by hean01-cendio** [comment]
@moobyfr Thanks for that input, just out of curiosity, does this work as expected using microsoft rdp client (mstsc) ?

This might be a bug that REALM is used and takes precedence over provided -d domain.

**Comment by hean01-cendio** [comment]
@moobyfr Could you try if the changes in PR #253 solves your sub domain problem, the PR changes the approach used to create service provider name from whats provided via command line to resolved fqdn of the server.

**Comment by moobyfr** [comment]
Hello,
I've tested the PR.
It's working as excepted: At launch rdesktop ask for the password, and then I'm logged in into the RDS server.
I can even omit the domain name in command line the REALM is used.
I tried with users from topdomain and subdomain.

**Comment by mgkuhn** [comment]
I read at https://docs.microsoft.com/en-us/windows/security/identity-protection/remote-credential-guard that since Windows Server 2016 and Windows 10 1607, Microsoft now fully supports proper Kerberos authentication and delegation. They call this feature “Windows Defender Remote Credential Guard”.

The prerequisites listed are:

 *  Must be running at least Windows 10, version 1607 or Windows Server 2016.
 *  Must allow Restricted Admin connections.
 *  Must allow the client’s domain user to access Remote Desktop connections.
 *  Must allow delegation of non-exportable credentials.

So the host must have in its Active Directory computer object the "trusted for delegation" flag set and also must have set a registry entry registry with

```
reg add HKLM\SYSTEM\CurrentControlSet\Control\Lsa /v DisableRestrictedAdmin /d 0 /t REG_DWORD
```

According to [a form answer here](https://social.msdn.microsoft.com/forums/en-US/da074f0f-0887-4151-88ea-19a671ed91d9/is-it-possible-to-do-true-kerberos-sso-using-rdp-?forum=os_windowsprotocols&prof=required):

> In [MS-CSSP], TSCredentials can now have a TSRemoteGuardCreds credential type.
> References:
> Protect Remote Desktop credentials with Remote Credential Guard
> https://docs.microsoft.com/en-us/windows/access-protection/remote-credential-guard
> [MS-CSSP]
> https://msdn.microsoft.com/en-us/library/cc226764.aspx
> [MS-RDPEAR]: Remote Desktop Protocol Authentication Redirection Virtual Channel
> https://msdn.microsoft.com/en-us/library/mt739840.aspx

Can anyone confirm that this also works with which rdesktop version?
